parquet-column/src/main/java/parquet/schema/Types.java

```diff
@@ -555,6 +607,342 @@ protected GroupType build(String name) {
+      public Q named(String name) {
+        if (parent instanceof  MapBuilder) {
+          ((MapBuilder) parent).setValueType(build("value"));
+          return (Q) ((MapBuilder) parent).named(name);
+        }
+
+        throw new UnsupportedOperationException("named not called on correct TypeBuilder object");
+      }
+
+      @Override
+      protected MapGroupValueBuilder<P, Q> self() {
+        return this;
+      }
+    }
+
+    protected void setKeyType(Type keyType) {
+      this.keyType = keyType;
+    }
+
+    protected void setValueType(Type valueType) {
+      this.valueType = valueType;
+    }
+
+    private Type keyType = null;
+    private Type valueType = null;
+    private Class<P> returnType = null;
+
+    public MapBuilder(P parent) {
+      super(parent);
+    }
+
+    private MapBuilder(Class<P> returnType) {
+      super(returnType);
+      this.returnType = returnType;
+    }
+
+    @Override
+    protected MapBuilder<P> self() {
+      return this;
+    }
+
+    public MapKeyBuilder<MapBuilder<P>, P> key(PrimitiveTypeName type) {
+      return new MapKeyBuilder<MapBuilder<P>, P>(this, parent, type);
+    }
+
+    public MapGroupKeyBuilder<MapBuilder<P>, P> groupKey() {
+      return new MapGroupKeyBuilder<MapBuilder<P>, P>(this, parent);
+    }
+
+    @Override
+    protected Type build(String name) {
+      if (keyType == null) {
+        keyType = STRING_KEY;
+      }
+      return buildGroup(repetition)
+          .as(OriginalType.MAP)
+          .addField(repeatedGroup()
+              .addFields(keyType, valueType)
+              .as(OriginalType.MAP_KEY_VALUE)
+              .named("map"))
+          .named(name);
+    }
+
+    public P named(String name) {
+      Preconditions.checkState(originalType == null,
+          "Cannot call \"as\" for a MAP type.");
+      return super.named(name);
+    }
+  }
+
+  public static class ListBuilder<P> extends Builder<ListBuilder<P>, P> {
+    private Type elementType = null;
+
+    public ListBuilder(P parent) {
+      super(parent);
+    }
+
+    public ListBuilder(Class<P> returnType) {
+      super(returnType);
+    }
+
+    @Override
+    protected ListBuilder<P> self() {
+      return this;
+    }
+
+    @Override
+    protected Type build(String name) {
+      Preconditions.checkNotNull(elementType, "List element type");
+      return buildGroup(repetition)
+          .as(OriginalType.LIST)
+          .addField(repeatedGroup()
+              .addFields(elementType)
+              .named("list"))
+          .named(name);
+    }
+
+    public P named(String name) {
+      Preconditions.checkState(originalType == null,
+          "Cannot call \"as\" for a LIST type.");
+      return super.named(name);
+    }
+
+    public GroupBuilder<ListBuilder<P>> group(Type.Repetition repetition) {
+      return (GroupBuilder<ListBuilder<P>>) new GroupBuilder<ListBuilder<P>>(this).repetition
+          (repetition);
+    }
+
+    public GroupBuilder<ListBuilder<P>> requiredGroup() {
```

Comment:
I like how the map builders use names like `requiredGroupValue` to call out that it is a single call. I think this should be similar and named `requiredGroupElement`.

While I'm thinking about it, these should ensure that they are called only once by validating that `elementType` is null. Similarly, the map builder should validate that `keyType` is null.